### PR TITLE
Basic BS adding trigger boilerplate

### PR DESCRIPTION
**Pull request by piekczyk**
# [Basic BS adding trigger boilerplate](https://app.shortcut.com/oazo-apps/story/3887/adding-trigger-from-basic-sell-form)

<please insert a shortcut link above>
  
## Changes 👷‍♀️
  <Please add short list of changes>
- added handling for adding basic sell trigger transaction
- added boilerplate necessary for farther development
  
## How to test 🧪
  <Please explain how to test your changes>
- use feature toggle BasicBS and try to add basic sell trigger (goerli testnet)
- test whether without BasicBS feature flag everything works properly

### CODE REVIEW

File: components/AppContext.ts [hunk 1]
```diff
@@ -78,6 +77,12 @@ import {
   AutomationChangeFeatureAction,
   automationChangeFeatureReducer,
 } from 'features/automation/protection/common/UITypes/AutomationFeatureChange'
+import {
+  BASIC_BUY_FORM_CHANGE,
+  BASIC_SELL_FORM_CHANGE,
+  BasicBSChangeAction,
+  basicBSFormChangeReducer,
```
Comment (johnnieskywalker) [inline components/AppContext.ts:84]: How about `OptimizationChangeAction`, `optimizationFormChangeReducer` ?

File: features/automation/useBasicSellStateInitializator.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    targetCollRatio,
+    maxBuyOrMinSellPrice,
+    continuous,
+    deviation,
+    isTriggerEnabled,
+  } = extractBasicBSData(autoTriggersData, type)
+  const collateralizationRatio = vault.collateralizationRatio.toNumber()
+
+  const publishKey = type === TriggerType.BasicBuy ? BASIC_BUY_FORM_CHANGE : BASIC_SELL_FORM_CHANGE
+  console.log(publishKey)
```
Comment (piotrkonowrocki) [inline features/automation/useBasicSellStateInitializator.ts:33]: `console` left

File: features/automation/useBasicSellStateInitializator.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import { useAppContext } from 'components/AppContextProvider'
+import { extractBasicBSData } from 'features/automation/common/basicBSTriggerData'
+import {
+  BASIC_BUY_FORM_CHANGE,
+  BASIC_SELL_FORM_CHANGE,
+} from 'features/automation/protection/common/UITypes/basicBSFormChange'
+import { TriggersData } from 'features/automation/protection/triggers/AutomationTriggersData'
+import { useEffect } from 'react'
+
+export function useBasicBSstateInitialization(
```
Comment (piotrkonowrocki) [inline features/automation/useBasicSellStateInitializator.ts:14]: This function is can return stuff for both basic buy and basic sell, but its name suggests that it concerns only basic sell. Do we have a name for grouping them? Also we would probably use it in the future for constant multiply?

File: components/vault/detailsSection/ContentCardTargetColRatio.tsx [hunk 4]
```diff
@@ -46,15 +44,14 @@ export function ContentCardTargetColRatio({
...
-  if (afterTargetColRatio && changeVariant)
+  if (afterTargetColRatio && !targetColRatio?.isEqualTo(afterTargetColRatio) && changeVariant)
     contentCardSettings.change = {
       value: `${formatted.afterTargetColRatio} ${t('system.cards.common.after')}`,
       variant: changeVariant,
     }
   if (threshold)
     contentCardSettings.footnote = t('auto-buy.continual-buy-threshold', {
       amount: formatted.threshold,
-      token,
```
[the remark this answers is not in the document]
Comment (piotrkonowrocki) [reply components/vault/detailsSection/ContentCardTargetColRatio.tsx:57]: Same question for `ContentCardTargetSellColRatio.tsx`.

File: features/automation/common/filterTriggersByType.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        triggerId: trigger.triggerId,
+        result: decodeTriggerData(trigger.commandAddress, networkId, trigger.executionParams),
+      }
+    })
+
+    return decodedTriggers.filter((decodedTrigger) =>
+      triggerTypes.includes(decodedTrigger.result[1]),
+    )
+  } catch (e) {
+    console.error(e)
```
Comment (piotrkonowrocki) [inline features/automation/common/filterTriggersByType.ts:23]: `console` left

File: features/automation/optimization/sidebars/SidebarSetupAutoBuy.tsx [hunk 6]
```diff
@@ -1,38 +1,55 @@
...
-  // const [activeAutomationFeature] = useObservable(
-  //   uiChanges.subscribe<AutomationChangeFeature>(AUTOMATION_CHANGE_FEATURE),
-  // )
+
+  const { uiChanges, txHelpers$ } = useAppContext()
+  const [txHelpers] = useObservable(txHelpers$)
+
   const [activeAutomationFeature] = useUIChanges<AutomationChangeFeature>(AUTOMATION_CHANGE_FEATURE)
+  const [uiState] = useUIChanges<BasicBSFormChange>(BASIC_BUY_FORM_CHANGE)
+  console.log(uiState)
```
Comment (piotrkonowrocki) [inline features/automation/optimization/sidebars/SidebarSetupAutoBuy.tsx:42]: `console` left

**Review by johnnieskywalker — CHANGES_REQUESTED** [review]
Just minor comments, and we're good to go

File: features/automation/optimization/sidebars/SidebarSetupAutoBuy.tsx [hunk 7]
```diff
@@ -73,13 +109,26 @@ export function SidebarSetupAutoBuy({
...
       ),
       primaryButton: {
         label: 'Confirm',
         disabled: false,
         action: () => {
-          addBasicBuyTriggerConfig.onClick(() => null)
+          if (txHelpers) {
+            txHelpers
+              .sendWithGasEstimation(addAutomationBotTrigger, txData)
+              .subscribe((next) => console.log(next))
```
Comment (piotrkonowrocki) [inline features/automation/optimization/sidebars/SidebarSetupAutoBuy.tsx:130]: `console` left

File: features/automation/protection/controls/sidebar/SidebarSetupAutoSell.tsx [hunk 8]
```diff
@@ -71,11 +117,27 @@ export function SidebarSetupAutoSell({
...
       ),
       primaryButton: {
         label: 'Confirm',
-        disabled: true,
+        disabled: false,
+        action: () => {
+          if (txHelpers) {
+            txHelpers
+              .sendWithGasEstimation(addAutomationBotTrigger, txData)
+              .subscribe((next) => console.log(next))
```
Comment (piotrkonowrocki) [inline features/automation/protection/controls/sidebar/SidebarSetupAutoSell.tsx:138]: `console`

**Review by piotrkonowrocki — CHANGES_REQUESTED** [review]
General comment: I left a comment in one place about naming something `...BS...`, when it actually was used in both BS and BB. I think it would be better to use a more universal name related to automation, because in the feature maybe we'll reuse it in constant multiply?

File: features/automation/protection/controls/sidebar/SidebarSetupAutoSell.tsx (hunk 8 above)
Comment (piekczyk) [reply features/automation/protection/controls/sidebar/SidebarSetupAutoSell.tsx:138]: its initial step, tx handling is a part of other story

File: features/automation/optimization/sidebars/SidebarSetupAutoBuy.tsx (hunk 7 above)
Comment (piekczyk) [reply features/automation/optimization/sidebars/SidebarSetupAutoBuy.tsx:130]: its initial step, tx handling is a part of other story

File: components/vault/detailsSection/ContentCardTargetColRatio.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (piekczyk) [reply components/vault/detailsSection/ContentCardTargetColRatio.tsx:57]: image
it doesn't make sense to add token when we are operating using token price

File: features/automation/useBasicSellStateInitializator.ts (hunk 2 above)
Comment (piekczyk) [reply features/automation/useBasicSellStateInitializator.ts]: removed

File: features/automation/common/filterTriggersByType.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  try {
+    const decodedTriggers = triggers.map((trigger) => {
+      return {
+        triggerId: trigger.triggerId,
+        result: decodeTriggerData(trigger.commandAddress, networkId, trigger.executionParams),
+      }
+    })
+
+    return decodedTriggers.filter((decodedTrigger) =>
+      triggerTypes.includes(decodedTrigger.result[1]),
```
[the remark this answers is not in the document]
Comment (piekczyk) [reply features/automation/common/filterTriggersByType.ts]: done

File: features/automation/common/basicBSTriggerData.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+  targetCollRatio: zero,
+  maxBuyOrMinSellPrice: zero,
+  continuous: false,
+  deviation: zero,
+  isTriggerEnabled: false,
+}
+
+export function extractBasicBSData(data: TriggersData, type: TriggerType): BasicBSTriggerData {
+  if (data.triggers && data.triggers.length > 0) {
+    const basicSellTriggers = getTriggersByType(data.triggers, [type])
```
[the remark this answers is not in the document]
Comment (piekczyk) [reply features/automation/common/basicBSTriggerData.ts:62]: currently extractBasicBSData is self explanatory as well

File: features/automation/common/filterTriggersByType.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
+import { decodeTriggerData, TriggerType } from '@oasisdex/automation'
+import { getNetworkId } from '@oasisdex/web3-context'
+import { TriggerRecord } from 'features/automation/protection/triggers/AutomationTriggersData'
+
+export function getTriggersByType(triggers: TriggerRecord[], triggerTypes: TriggerType[]) {
+  const MAINNET_ID = 1
+  const GOERLI_ID = 5
```
[the remark this answers is not in the document]
Comment (piekczyk) [reply features/automation/common/filterTriggersByType.ts:7]: done

File: features/automation/optimization/sidebars/SidebarSetupAutoBuy.tsx (hunk 6 above)
Comment (piekczyk) [reply features/automation/optimization/sidebars/SidebarSetupAutoBuy.tsx:42]: removed

**Review by johnnieskywalker — APPROVED** [review]
gj

**Pull request merged**